artifact.go

```diff
@@ -215,55 +218,413 @@ func (m artifactModule) postBuildArtifactHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	files, ok := parseMultipartFormData(c, buildID)
+	if !ok {
+		return
+	}
+
+	artifacts, ok := m.createArtifacts(c, files, buildID)
+	if !ok {
+		return
+	}
+
+	summaries := make([]TestResultSummary, 0, len(artifacts))
+	lotsOfDetails := make([]TestResultDetail, 0)
+
+	for _, artifact := range artifacts {
+		summary, details, err := getTestSummaryAndDetails(artifact.Data, artifact.ArtifactID, buildID)
+		if err != nil {
+			log.Warn().
+				WithError(err).
+				WithString("filename", artifact.Name).
+				WithUint("build", buildID).
+				WithUint("artifact", artifact.ArtifactID).
+				Message("Failed to unmarshal; invalid TRX/XML format, or the structure" +
+					" receiving the data is malformed.")
+
+			ginutil.WriteProblemError(c, err,
+				problem.Response{
+					Type:   "/prob/unexpected-response-format",
+					Status: 502,
+					Title:  "Unexpected response format.",
+					Detail: fmt.Sprintf(
+						"Failed parsing test result ID %d, for build with ID %d in"+
+							" database. Expected TRX/XML format. Could also be due to an"+
+							" internal bug.", summary.ArtifactID, buildID),
+				})
+			return
+		}
+
+		summary.FileName = artifact.FileName
+		summaries = append(summaries, summary)
+		lotsOfDetails = append(lotsOfDetails, details...)
+	}
+
+	summaryList := TestResultListSummary{
+		Summaries: summaries,
+		BuildID:   buildID}
+	for _, summary := range summaryList.Summaries {
+		err := m.Database.
+			Create(summary).
+			Error
+		if err != nil {
+			ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+				"Failed saving test result summary from artifact with ID %d, for build"+
+					" with ID %d in database.",
+				summary.ArtifactID, buildID))
+			return
+		}
+
+		summaryList.Failed += summary.Failed
+		summaryList.Passed += summary.Passed
+		summaryList.Skipped += summary.Skipped
+	}
+
+	summaryList.Total =
+		summaryList.Failed +
+			summaryList.Passed +
+			summaryList.Skipped
+
+	err := m.Database.
+		CreateInBatches(lotsOfDetails, 100).
+		Error
+	if err != nil {
+		ginutil.WriteDBWriteError(c, err, fmt.Sprintf(
+			"Failed saving test result details for build with ID %d in database.",
+			buildID))
+	}
+
+	c.JSON(http.StatusOK, summaryList)
+}
+
+// getBuildAllTestResultDetailsHandler godoc
+// @summary Get all test result details for specified build
+// @tags artifact
+// @param buildid path int true "Build ID"
+// @success 200 {object} []TestResultDetail
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result-details [get]
+func (m artifactModule) getBuildAllTestResultDetailsHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	details := []TestResultDetail{}
+	err := m.Database.
+		Where(&TestResultDetail{BuildID: uint(buildID)}).
```

Comment:
That cast seems to be unneeded.

```suggestion
		Where(&TestResultDetail{BuildID: buildID}).
```